sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,768 @@
+      if (getDataSourceProviderFn() instanceof HasDisplayData) {
+        ((HasDisplayData) getDataSourceProviderFn()).populateDisplayData(builder);
+      }
+    }
+  }
+
+  /**
+   * A POJO describing a {@link DataSource}, providing all properties allowing to create a {@link
+   * DataSource}.
+   */
+  @AutoValue
+  public abstract static class DataSourceConfiguration implements Serializable {
+    @Nullable
+    public abstract String getUrl();
+
+    @Nullable
+    public abstract String getUsername();
+
+    @Nullable
+    public abstract String getPassword();
+
+    @Nullable
+    public abstract PrivateKey getPrivateKey();
+
+    @Nullable
+    public abstract String getOauthToken();
+
+    @Nullable
+    public abstract String getDatabase();
+
+    @Nullable
+    public abstract String getWarehouse();
+
+    @Nullable
+    public abstract String getSchema();
+
+    @Nullable
+    public abstract String getServerName();
+
+    @Nullable
+    public abstract Integer getPortNumber();
+
+    @Nullable
+    public abstract String getRole();
+
+    @Nullable
+    public abstract Integer getLoginTimeout();
+
+    @Nullable
+    public abstract Boolean getSsl();
+
+    @Nullable
+    public abstract DataSource getDataSource();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setUrl(String url);
+
+      abstract Builder setUsername(String username);
+
+      abstract Builder setPassword(String password);
+
+      abstract Builder setPrivateKey(PrivateKey privateKey);
+
+      abstract Builder setOauthToken(String oauthToken);
+
+      abstract Builder setDatabase(String database);
+
+      abstract Builder setWarehouse(String warehouse);
+
+      abstract Builder setSchema(String schema);
+
+      abstract Builder setServerName(String serverName);
+
+      abstract Builder setPortNumber(Integer portNumber);
+
+      abstract Builder setRole(String role);
+
+      abstract Builder setLoginTimeout(Integer loginTimeout);
+
+      abstract Builder setSsl(Boolean ssl);
+
+      abstract Builder setDataSource(DataSource dataSource);
+
+      abstract DataSourceConfiguration build();
+    }
+
+    /**
+     * Creates {@link DataSourceConfiguration} from existing instance of {@link DataSource}.
+     *
+     * @param dataSource - an instance of {@link DataSource}.
+     */
+    public static DataSourceConfiguration create(DataSource dataSource) {
+      checkArgument(dataSource instanceof Serializable, "dataSource must be Serializable");
+      return new AutoValue_SnowflakeIO_DataSourceConfiguration.Builder()
+          .setDataSource(dataSource)
+          .build();
+    }
+
+    /**
+     * Creates {@link DataSourceConfiguration} from instance of {@link SnowflakeCredentials}.
+     *
+     * @param credentials - an instance of {@link SnowflakeCredentials}.
+     */
+    public static DataSourceConfiguration create(SnowflakeCredentials credentials) {
```

Comment:
Can we introduce just one create(SnowflakeCredentials credentials) method and fork based on the type of object passed in ?